### PR TITLE
Navigation Component: Search Control in Menu Titles

### LINKED ISSUE #25252: Navigation Component: Search
Add a mechanism to add search/filtering capabilities to the `NavigationMenu` titles.

https://github.com/WordPress/gutenberg/blob/668a710b3eb9c942b21eb5913fbe304b15ea71dd/packages/components/src/navigation/menu.js#L52-L60

Example:

Add a lens icon that toggles between title and search box.
The search box could filter the `NavigationMenu` children, or perform a search via API.

([Source](https://github.com/WordPress/gutenberg/issues/23939#issuecomment-662012836))

### PR DESCRIPTION
**Pull request by Copons**
## Description

Introduce a new `NavigationMenuTitle` component that can optionally become a search box, filtering the menu items.

Fixes #25252

#### Things to do

- [x] The search uses the `appear` animation instead of `slide-in`. This is because the parent component uses `slide-in`, and for some reasons whenever I toggle the search box, both animations would be triggered. (Solved: it was caused by the input gaining focus during the animation, which scrolled it into view, messing up the animation.)
- [x] ~My intention would be to make the input field to look like the inserter search box, which is heavily customized and requires more involvement than I initially thought.~ (Not worthy: a normal `TextControl` will do just fine)
- [ ] There are a lot of magic numbers in the CSS, I'm aware of it. 🙂 
- [x] ~Maybe externalize the `normalize` function, also used by the inserter search.~ (It has a very low footprint and it's used by several other components already. We might want to create a general utility for it, but this is not the right place.)
- [x] Gracefully handle empty `NavigationGroup`.
- [x] Add an `onSearch` function to handle custom search logic (this might require its own example in the storybook).
- [x] ~Explore if there's a simple way to search inside items rendering custom components.~ (Discussion moved to #25509)

## How has this been tested?

- `npm run storybook:dev`
- Head over to Components / Navigation.
- The "Home" menu is now searchable.

## Screenshots 

Sep-23-2020 18-58-38

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by Copons** [comment]
@david-szabo97 Let me loop you in on what I'm proposing here, which contains some changes on the navigation tree structure.

tl;dr: we have the tree, let's use it! 😄 

### Item `_isVisible` instead of menu `isActive`

With the search, items can be hidden because of 2 reasons: 1) the belong to an inactive menu; 2) they are filtered out by the search.
This means that `NavigationMenuContext`'s `isActive` wasn't enough anymore, and so I've removed it altogether.

In its stead, I've added a new `_isVisible` property to the [navigation tree's items](https://github.com/WordPress/gutenberg/pull/25315/files#diff-d481e74153239e32660d94288e855599).
Now the nav tree items update whenever `activeMenu` or `search` (the search field state) change.

This property (accessible for example with `navigationTree.getItem( item )?._isVisible`) includes by default the active menu check, and adds the search filtering as well.

### Implementation

`NavigationItem` uses it directly to determine if the item should be rendered or not.

`NavigationMenu` replaces the `isActive` context with a direct `activeMenu !== menu` check.

[`NavigationGroup`](https://github.com/WordPress/gutenberg/pull/25315/files#diff-2247041bd3b4c4f8941d0e650d295cbc) is more complicated.
We don't have it in the tree, and since it's almost entirely a presentational wrapper, I'd rather keep it out unless really unavoidable.
Loop with `Children.forEach`: if all its children are marked as not `_isVisible` in the tree, then the group is empty. It doesn't need to render anything, but it can just pass-through children (which are all "invisible" items, so they eventually won't be rendered at all).

_Note: apparently there's no way to break from a `Children.forEach`. It's a bummer, but probably not a big deal? We could turn it into an array with `Children.toArray` and use a normal loop, but I guess that would negate the perf gain anyway._ 🤷

**Comment by david-szabo97** [comment]
Looks good to me! Any reason to use `_` prefix? It's a computed property but I don't think it should be a marked as private (if that was the intention).

> apparently there's no way to break from a Children.forEach.
Too bad, probably it's still cheaper to loop through than `toArray`.

**Comment by david-szabo97** [comment]
> Explore if there's a simple way to search inside items rendering custom components.

This looks promising: https://github.com/fernandopasik/react-children-utilities/blob/master/docs/only-text.md
Maybe we should replicate this to achieve searching in custom components.

**Comment by Copons** [comment]
> Looks good to me! Any reason to use `_` prefix? It's a computed property but I don't think it should be a marked as private (if that was the intention).

It was both to distinguish it from the components props, and to avoid clashing with a hypothetical `isVisible` prop.

**Comment by david-szabo97** [comment]
> Just throwing out a random idea: what if we had a keywords prop, similar in function to the blocks' keywords?

Well, this is the least expensive way to do it. I'm not sure if it's developer friendly though.  It's a good starting point and we might need this for something else as well. (NavigationItem's label might not be the best describing text for searching in every case)

**Comment by Copons** [comment]
> > Just throwing out a random idea: what if we had a keywords prop, similar in function to the blocks' keywords?
> 
> Well, this is the least expensive way to do it. I'm not sure if it's developer friendly though. It's a good starting point and we might need this for something else as well. (NavigationItem's label might not be the best describing text for searching in every case)

Yup.
I think it's worth discussing it separately, as it's a decision we would eventually need to stick to indefinitely.
I'll open an issue for that, so that we could wrap this up with just the title search. 🙂

**Comment by shaunandrews** [comment]
--

image

I noticed the focus outline gets cutoff for the search button.

--

As I'm searching, I expected to be able to use the up/down arrow keys to navigate the search results.

--

I'm able to use the `tab` key to move focus to the `x` button to close the search, but when it closes my keyboard's focus disappears; I'd expect it to return to the search button after closing the input.

**Comment by Copons** [comment]
> I think the icon should be white, not blue.

I took the liberty of making it grey, white on hover, to be consistent with the other buttons of the menu.
I'm aware that the draft has "inverted" colors (links are white, while normal text is grey), but we've missed it when we merged the initial PR.
This is not really the place for inverting all colors, but we'll take care of it eventually.

> I noticed the focus outline gets cutoff for the search button.
>
> I'm able to use the `tab` key to move focus to the `x` button to close the search, but when it closes my keyboard's focus disappears; I'd expect it to return to the search button after closing the input.
>
> I expect to be able to hit the `esc` button to close the search.

✅ 

> As I'm searching, I expected to be able to use the up/down arrow keys to navigate the search results.

✅ (It's not limited to the search, but the menu itself is now navigable via up/down arrows.)

The behaviour with the search would be:
- Start with focus on the input.
- TAB -> focus on the close search button.
- TAB -> focus on the first item of the menu.
- UP/DOWN -> cycle through the menu items.
- TAB would work too, but it "breaks out" of the cycle either when TABbing out of the last item, or when SHIFT+TABbing on the first item (focus back on the close search button).

This seems consistent with the Inserter.

Note: I've realized that the `Navigation` focus is really a bit messy, I'll try to take care of it ASAP.
I won't pretend to know the a11y implications, but I can at least try to make the keyboard navigation feel good!

**Comment by shaunandrews** [comment]
> we'll take care of it eventually.

For sure.

> The behaviour with the search would be:
> 
> Start with focus on the input.
> TAB -> focus on the close search button.
> TAB -> focus on the first item of the menu.
> UP/DOWN -> cycle through the menu items.
> TAB would work too, but it "breaks out" of the cycle either when TABbing out of the last item, or when SHIFT+TABbing on the first item (focus back on the close search button).

I think it might be better if you could use the DOWN key while the focus is on the input. Otherwise, this feels like a good start to me. Once we have some real content in the list, I imagine we'll be in a better place to do a more thorough review of the interactions.

**Review by shaunandrews — APPROVED** [review]
I'm ignoring all those magic CSS values ;)

**Comment by enriquesanchez** [comment]
> The behaviour with the search would be:
> 
> Start with focus on the input.
> TAB -> focus on the close search button.
> TAB -> focus on the first item of the menu.
> UP/DOWN -> cycle through the menu items.
> TAB would work too, but it "breaks out" of the cycle either when TABbing out of the last item, or when SHIFT+TABbing on the first item (focus back on the close search button).

The keyboard navigation feels right to me 👍. I don't think we should allow to use the <kbd>Down</kbd> arrow key while the focus is on the input and agree it should behave like the inserter.

---

Something else I noticed is that the `Search` and `X` buttons are missing an `aria-label`. Without it, they're currently just announced as 'buttons' to screen readers without any more context. The `aria-label` should be descriptive of what the action will do, for example: 'Search posts'. 

Similar to how it's done in the inserter, it's also important to announce the number of search results found.

I'm not sure if these search results will display more than one group of items, but if they do, it's also a good idea to look at the inserter for guidance on how to communicate this information to assistive technology. We use `role=listbox` and an `aria-label` on each group of results.

**Comment by Copons** [comment]
> I don't think we should allow to use the Down arrow key while the focus is on the input

I agree with this, but wouldn't object to a good reasoning for "intercepting" the arrow keys.

I'm aware that lots of search inputs use up/down to navigate the search suggestion, but at the same time when I tested the search, being able to move the text cursor with them felt right.

It's probably because the Nav filtering doesn't look like an autocomplete/suggestion box, and the results are visually separated from the search input.
Maybe a visually impaired user would have the complete opposite feeling.

**Comment by Copons** [comment]
> I'm ignoring all those magic CSS values ;)

@shaunandrews I've replaced most magic values with the `space` function (I somehow missed that before!), and added inline comments to the others.

I'll open an issue to remind us to use `space` on all the other styled components.

> Similar to how it's done in the inserter, it's also important to announce the number of search results found.

@enriquesanchez Nice one! TIL about the `withSpokenMessages` HOC! ❤️ 

> I'm not sure if these search results will display more than one group of items, but if they do, it's also a good idea to look at the inserter for guidance on how to communicate this information to assistive technology. We use `role=listbox` and an `aria-label` on each group of results.

Menus can be optionally split in groups (`NavigateGroup`, which is basically a nested list with a title) and these groups will remain even when searching (empty groups will be automatically hidden).
I'm not exactly sure what kind of information we should add and where.
Is something like this what you have in mind?

```jsx
// NavigationMenu
<h2 id="menuId">Menu Title</h2>
<ul role="listbox" aria-labelledby="menuId"> ... </ul>

// NavigationGroup
<li>
  <h3 id="groupId">Group Title</h3>
  <ul role="listbox" aria-labelledby="groupId"> ... </ul>
</li>
```

As far as I can understand, what you're asking is unrelated to the search function, so I'll open a new issue for it.

File: packages/components/src/navigation/item/use-navigation-tree-item.js [hunk 1]
```diff
@@ -8,19 +8,29 @@ import { useEffect } from '@wordpress/element';
...
 	} = useNavigationContext();
-	const { menu } = useNavigationMenuContext();
+	const { menu, search } = useNavigationMenuContext();
 
 	const key = props.item;
 	useEffect( () => {
-		addItem( key, { ...props, menu } );
+		const isMenuActive = activeMenu === menu;
+		const isItemVisible =
+			! search || normalizedSearch( props.title, search );
```
Comment (vindl) [inline packages/components/src/navigation/item/use-navigation-tree-item.js:24]: I think this is fine for now, but I think we'll need to think of how to search from external dataset whose items are not completely included in navigation menu? For example, we might have Pages navigation with thousands of items, and we might not want to render them all within the menu.

Comment (Copons) [reply packages/components/src/navigation/item/use-navigation-tree-item.js:24]: There is an `onSearch` prop that can be used to handle this kinds of things, e.g. to perform an actual database search.
(At least that's the idea, but I haven't tested it in a real-life scenario)

**Comment by shaunandrews** [comment]
> I've replaced most magic values with the space function (I somehow missed that before!), and added inline comments to the others.

> I'll open an issue to remind us to use space on all the other styled components.

This is actually the first I've heard of the `space` function myself. I'm a little hesitant, as it seems to completely ignore the defined CSS variables from `base-styles`: https://github.com/WordPress/gutenberg/blob/a63cc627fa9de3ba9cfc3fcdf5ac07e0549dd166/packages/base-styles/_variables.scss#L27

I'll admit, I'm unsure which is the preferred way to output the grid values but a quick search shows that the `space` function is only used in one or two places.

**Comment by enriquesanchez** [comment]
>Is something like this what you have in mind?

**Comment by Copons** [comment]
@enriquesanchez Apologize for the long wait, I've been AFK last week.

I've pushed a couple of commits adding:
- `role="group"` to groups.
- `role="listitem"` to items.
- `aria-labelledby` to both menus and groups, pointing to their titles.

Some possible concerns:

- Given that `role="list"` is discouraged for `<ul>`, only groups have a role, but **all** items (both grouped and otherwise) are `role="listitem"`.
This is because of the technical implementation (items are the same inside or outside groups).
So, items in menus without groups will be `listitem` even if they don't belong to a `list` or `group`. Is this an a11y issue?

- Searchable menus swap their title with a search bar, which means that their `aria-labelledby` needs to be a little more involved.
My solution is to use the same ID for both the title (when the search is off) and the input label (when the search is on). Those elements can't be rendered at the same time, so there won't be duplicate IDs.
Though, I do wonder if changing the ARIA label is an a11y issue ("Menu Name" -> "Search in Menu Name").

**Comment by enriquesanchez** [comment]
Hi @Copons! 👋

I just gave the updte a test and I think it worked really well. The groups were communicated properly and I was able to search and navigate the list of results.

>Given that role="list" is discouraged for <ul>, only groups have a role, but all items (both grouped and otherwise) are role="listitem".

I'm still a bit uncertain about this, and I'm left wondeting if we need `role="listitem"`at all, particularly because we're working with semantic markup (`<ul>` and `<li>`).

Would it be too much trouble to remove `role="listitem"` from the `<li>` and give it another test? I'd like to make sure if there's any change in behavior for screen readers.

>Searchable menus swap their title with a search bar, which means that their aria-labelledby needs to be a little more involved.

I did not find any issues with this and I think your approach makes sense 👍

--- 

I also want to ask because I couldn't figure out how, but is there a way I can perform a search that renders more than one group? On my test I was only able to see one group at a time.

**Comment by Copons** [comment]
> Would it be too much trouble to remove `role="listitem"` from the `<li>` and give it another test? I'd like to make sure if there's any change in behavior for screen readers.

@enriquesanchez No trouble at all! I've pushed a commit removing it, feel free to try what happens! 🙂 

> I also want to ask because I couldn't figure out how, but is there a way I can perform a search that renders more than one group? On my test I was only able to see one group at a time.

It's actually very easy, but not very clear given that the second group only contains one item ("Controlled Search").
It's enough to type one or two characters included in "Controlled Search" that would also match some items from the first group.
E.g. "co" would match "Corge" in the first group and "Controlled Search" in the second.
Even just typing "r" matches several items from the first group, as well as "Controlled Search".

**Comment by enriquesanchez** [comment]
Hey @Copons 

Thank you for working on those changes. Tested again and I couldn't notice a difference. Both groups and items under them are communicated properly. 

I think this is in good shape 👍

**Comment by Copons** [comment]
@enriquesanchez Thanks for looking into this!

I need to put it back into WIP because it doesn't play well with some updates we did in the site editor, but I'll make sure to keep all the a11y improvements we introduced! ✨

**Comment by mattwiebe** [comment]
I was a bit confused because I didn't read the testing directions properly and was trying to figure out why the search wasn't showing up in the Site Editor sidebar.

But now that I tested in Storybook, it looks and works well for me. I wasn't sure if it would search only within the subsection (Controlled Search) but the placeholder adapts nicely to that ("Search in Controlled Search"), which is a nice touch.

But coming back to my initial confusion, are we intending to use the search on the Site Editor sidebar?

File: packages/components/src/navigation/menu/menu-title.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	const onCloseSearch = () => {
+		setIsSearching( false );
+
+		// Wait for the slide-in animation to complete before focusing the search button.
+		// eslint-disable-next-line @wordpress/react-no-unsafe-timeout
+		setTimeout( () => {
+			searchButtonRef.current.focus();
+		}, 100 );
```
Comment (david-szabo97) [inline packages/components/src/navigation/menu/menu-title.js:38]: This is very minor, but I see we are using this `100` magic number a few times. Maybe worth a constant?

File: packages/components/src/navigation/utils.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/**
+ * External dependencies
+ */
+import { deburr } from 'lodash';
+
+// @see packages/block-editor/src/components/inserter/search-items.js
```
Comment (david-szabo97) [inline packages/components/src/navigation/utils.js:6]: Would it worth opening an issue to extract this as a util function? So we can reuse it in both places.

**Comment by david-szabo97** [comment]
Moving into a submenu, the focus is lost. I'm not sure what would be the best a11y-wise. Maybe we should focus on the first item or the back button? 🤔

File: packages/components/src/navigation/utils.js (hunk 3 above)
Comment (Copons) [reply packages/components/src/navigation/utils.js:6]: Yes definitely!
My intention was to open a follow up right after merging this, and this is a nice reminder, since at this point I'm sure I'd have forgot about it. 😄

**Comment by Copons** [comment]
> But coming back to my initial confusion, are we intending to use the search on the Site Editor sidebar?

@mattwiebe Yes.
There are at least two locations where it would be useful:
- Search for templates, especially when we'll remove the "All templates" menu, currently used for testing purposes.
- Search for content. When we moved the content switcher from the header toolbar, we lost the "search for post" functionality (because its component's style was heavily clashing with the sidebar). That imho would be a nice way to test this built-in search in a real case scenario.

> Moving into a submenu, the focus is lost. I'm not sure what would be the best a11y-wise. Maybe we should focus on the first item or the back button? 🤔
> 
> EDIT: I don't think this is a blocker. Fine in a separate PR so we can merge this and get the ball rolling.

@david-szabo97 I see what you mean... 🤔 
The focus is indeed lost, but a single <kbd>tab</kbd> hit would move it to the right place (the top of the sidebar, which in this case would be the "back to parent menu" button).

Probably not a big deal, and not really related to the search function. 🙂

**Review by vindl — APPROVED** [review]
This is testing fine for me and the code looks good! 🚀

**Review by david-szabo97 — APPROVED** [review]
